frontend/docsettings.lua

```diff
@@ -244,9 +286,20 @@ function DocSettings:flush(data)
                 ffiutil.fsyncDirectory(sidecar_file)
             end
 
+            -- move cover file to the metadata file location
+            if not no_cover then
+                if self.cover_file == nil then
+                    self.cover_file = self:getCustomBookCover(self.data.doc_path) or false
+                end
+                if self.cover_file and util.splitFilePathName(self.cover_file) ~= sidecar_dir then
```

Comment:
Why have this only here in :flush(), and why not have it more generic like I suggested at https://github.com/koreader/koreader/pull/10329#discussion_r1169194161 ?
(I would rename your existing `DocSettings:getCoverFile(sidecar_dir)` to `:_findCoverFile(sidecar_dir)` to get its name for that more clever caching new method.)

Reply:
Do you suggest to get and cache cover_file in `open()`?
As we discussed, `doc_settings` are frequently opened without flush so no need to have a cover file.
I made it in `flush()` to be used in scheduled doc_settings saving only.

Reply:
> Do you suggest to get and cache cover_file in open()?

No, I suggest you don't do anything special in open()/flush()...
You just have and use `cover_file = DocSettings:getCoverFile()` only when you need a `cover_file` (including in flush(), to see if you should move it) - and this new getCoverFile() will ensure the caching, without open/flush() knowing/having to care.

Reply:
Sorry, don't understand.
I try to handle all possible cases like : doc_settings and cover_file in different folders; sdr folder location has been changed and a book has old sdr location; cover_file is added while a book is opened (including opened for the first time) and so on.
So I request cover_file (that could be changed at any moment) every time flushing is done.

Reply:
Yes, you should continue to do that here in :flush().
But the snippet I commented on should read:
```diff
            if not no_cover then
-                if self.cover_file == nil then
-                    self.cover_file = self:getCustomBookCover(self.data.doc_path) or false
-                end
-                if self.cover_file and util.splitFilePathName(self.cover_file
+                local cover_file = self:getCustomBookCover(self.data.doc_path)
+                if cover_file and util.splitFilePathName(self.cover_file
```
that is: handle the cache (cached as self._cover_file) only in `self:getCustomBookCover()` (either extend your existing `self:getCustomBookCover()` to do that, or add another function `self:getCoverFile()` that will return the cache, or call `self:getCustomBookCover()` and cache it - I prefer the latter, as `local cover_file =self:getCoverFile()` just reads as nice English :).
So that if it was fetched when showing BookInfo, or browsing CoverBrowser before the first :flush(), this first :flush() will get it from the cache without even knowing.

Reply:
And this cache must be updated in `BookInfo:setCustomBookCover()`, in `DocSettings:purge()`, in `DocSettings:update()`; and in `DocSettings:flush()` itself (if we have changed the sdr location while a book is opened).
Isn't it too much?

Reply:
If not "updated", just "reset/trashed" - so it's only rechecked when later needed.
In these parts where you know the cache is no longer valid, juste call another method to clear it (or self:getCoverFile(false)), or just set `self._cover_file = nil`.